src/resources/dart-3-migration.md

```diff
@@ -0,0 +1,545 @@
+
+```terminal
+$ dart pub get
+
+Because pkg1 doesn't support null safety, version solving failed.
+The lower bound of "sdk: '>=2.9.0 <3.0.0'" must be 2.12.0 or higher to enable null safety.
+```
+
+Libraries that opt out of null safety with [language version comments](/guides/language/evolution#per-library-language-version-selection)
+that select any language version below `2.12` will cause analysis or compilation
+errors:
+
+```terminal
+$ dart analyze .
+Analyzing ....                         0.6s
+
+  error • lib/pkg1.dart:1:1 • The language version must be >=2.12.0. 
+  Try removing the language version override and migrating the code.
+  • illegal_language_version_override
+```
+
+```terminal
+$ dart run bin/my_app.dart
+../pkg1/lib/pkg1.dart:1:1: Error: Library doesn't support null safety.
+// @dart=2.9
+^^^^^^^^^^^^
+```
+
+#### Migration
+
+Before beginning any migration to Dart 3, ensure your app or package has been 100% migrated to enable null safety. 
+This requires a Dart `2.19` SDK, not a Dart 3 SDK. For details, see the [null safety migration guide](/null-safety/migration-guide).
+
+### Colon-syntax for default values
+
+For historical reasons, named optional parameters could specify their default
+value using either `:` or `=`. In Dart 3, only the `=` syntax is allowed.
+
+#### Scope
+
+This is an [*unversioned* change](#unversioned-vs-versioned-changes), that applies to all Dart 3 code.
+
+#### Symptom
+
+Dart analysis produces errors like:
+
+```
+line 2 • Using a colon as a separator before a default value is no longer supported.
+```
+
+#### Migration
+
+Change from using colons:
+
+```dart
+int someInt({int x: 0}) => x;
+```
+
+To using equals:
+
+```dart
+int someInt({int x = 0}) => x;
+```
+
+This migration can be made manually, or automated with `dart fix`:
+
+```
+$ dart fix --apply --code=obsolete_colon_for_default_value
+```
+
+### `mixin`
+
+Pre-Dart 3, any `class` could be used as a `mixin`,
+as long as it had no declared constructors and no superclass other than `Object`.
+
+In Dart 3, classes declared in libraries at language version 3.0 or later
+can't be used as mixins unless marked `mixin`.
+This restriction applies to code in any library attempting to use the class as a mixin, 
+regardless of the latter library's language version.
+
+#### Scope
+
+This is a [*versioned* change](#unversioned-vs-versioned-changes), that only applies to language version 3.0 or later.
+
+#### Symptom
+
+An analysis error like:
+
+```
+Mixin can only be applied to class.
+```
+
+The analyzer produces this diagnostic when a class that is neither a
+`mixin class` nor a `mixin` is used in a `with` clause.
+
+#### Migration
+
+Determine if the class is intended to be used as a mixin.
+
+If the class defines an interface, consider using `implements`.
+
+### `switch`
+
+Dart 3.0 interprets [switch cases](/language/control-flow#switch-and-case)
+as patterns [*TODO* LINK] instead of constant expressions. 
```

Comment:
There's still a TODO link here.

I have already made a note for myself to come back to this page and add links to the first sentence (where it says "records, patterns, and class modifiers") after we actually publish everything early tomorrow morning (~7AM PST). I can make sure to add the link here if you want to remove this TODO from the text for now.

```suggestion
as patterns instead of constant expressions. 
```